src/go.ts

```diff
@@ -0,0 +1,298 @@
+      utils.shell(`git checkout -b ${this.gitBranch}`);
+    }
+
+    this.syncModules(repoDir);
+    try {
+      utils.shell('git add .');
+      utils.shell('git diff-index --exit-code HEAD --');
+      console.log('No changes. Skipping release');
+      return {};
+    } catch (err) {
+      // changes exist, thats ok.
+    }
+
+    const commitMessage = process.env.GIT_COMMIT_MESSAGE ?? this.createReleaseMessage(modules);
+
+    utils.shell(`git config user.name ${this.gitUsername}`);
+    utils.shell(`git config user.email ${this.gitUseremail}`);
+    utils.shell(`git commit -m "${commitMessage}"`);
+
+    const tags = modules.map(m => this.createTag(m));
+
+    if (this.dryRun) {
+      console.log(`Will push to branch: ${this.gitBranch}`);
+      tags.forEach(t => console.log(`Will push tag: ${t}`));
+    } else {
+      utils.shell(`git push origin ${this.gitBranch}`);
+      tags.forEach(t => utils.shell(`git push origin ${t}`));
+    }
+    return { tags, commitMessage };
+  }
+
+  private collectModules(dir: string): GoModule[] {
+    const modules: GoModule[] = [];
+
+    for (const p of [...fs.readdirSync(dir), '.']) {
+      const modFile = path.join(dir, p, 'go.mod');
+      if (fs.existsSync(modFile)) {
+        modules.push(this.createModule(modFile));
+      }
+    }
+    return modules;
+  }
+
+  private createModule(modFile: string): GoModule {
+
+    const version = this.extractVersion(path.dirname(modFile));
+    const majorVersion = version.split('.')[0];
+
+    const cannonicalNameParts = [];
+    for (const line of fs.readFileSync(modFile).toString().split('\n')) {
+      if (line.startsWith('module ')) {
+        cannonicalNameParts.push(...line.split(' ')[1].split('/'));
+        break;
+      }
+    }
+
+    if (cannonicalNameParts.length === 0) {
+      throw new Error(`Unable to detected module declaration in ${modFile}`);
+    }
+
+    const repoURL = cannonicalNameParts.slice(0, 3).join('/');
+
+    if (!repoURL.startsWith('github.com')) {
+      throw new Error(`Repository must be hosted on github.com. Found: ${repoURL}`);
+    }
+
+    const cannonicalName = cannonicalNameParts.join('/');
+    const repoPath = cannonicalNameParts.slice(3).join('/').replace(`/v${majorVersion}`, '');
+
+    return { modFile, version, cannonicalName, repoPath, repoURL };
+
+  }
+
+  private syncModules(repoDir: string) {
+    const topLevel = path.join(repoDir, 'go.mod');
+    if (fs.existsSync(topLevel)) {
+      // with top level modules we sync the entire repository
+      utils.removeDirectory(repoDir, { includeRoot: false, exclude: ['.git'] });
+    } else {
+      // otherwise, we selectively remove the submodules only.
+      for (const p of fs.readdirSync(repoDir)) {
+        const submodule = path.join(repoDir, p, 'go.mod');
+        if (fs.existsSync(submodule)) {
+          utils.removeDirectory(path.join(repoDir, p));
+        }
+      }
+    }
+    utils.shell(`cp -r ${this.dir}/* ${repoDir}`, { shell: true });
```

Comment:
Why not use fs-extra copy?

Reply:
I think I developed an inherent aversion to dependencies. But no good reason here. I'll add it - its useful for other stuff too.